File: heron/common/src/java/org/apache/heron/common/basics/FileUtils.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.heron.common.basics;

import java.io.File;
import java.io.IOException;
import java.nio.file.DirectoryNotEmptyException;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.OpenOption;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Utilities related to File.
 */
public final class FileUtils {
  private static final Logger LOG = Logger.getLogger(FileUtils.class.getName());

  private FileUtils() {
  }

  public static boolean deleteFile(String filename) {
    Path file = new File(filename).toPath();
    try {
      Files.delete(file);
    } catch (NoSuchFileException x) {
      LOG.severe("file does not exist: " + file);
      return false;
    } catch (DirectoryNotEmptyException x) {
      LOG.severe("Path is an not empty directory: " + file);
      return false;
    } catch (IOException x) {
      // File permission problems are caught here.
      LOG.log(Level.SEVERE, "Failed to delete file due to unexpected exception:", x);
      return false;
    }

    return true;
  }

  public static boolean copyFile(String source, String target) {
    Path sourcePath = new File(source).toPath();
    Path targetPath = new File(target).toPath();
    try {
      Files.copy(sourcePath, targetPath, StandardCopyOption.REPLACE_EXISTING);
    } catch (IOException e) {
      LOG.log(Level.SEVERE, "Failed to copy file from " + source + " to target: " + target, e);
      return false;
    }

    return true;
  }

  public static boolean writeToFile(String filename, byte[] contents, boolean overwrite) {
    // default Files behavior is to overwrite. If we specify no overwrite then CREATE_NEW fails
    // if the file exist. This operation is atomic.
    OpenOption[] options = overwrite
        ? new OpenOption[]{} : new OpenOption[]{StandardOpenOption.CREATE_NEW};

    try {
      Files.write(new File(filename).toPath(), contents, options);
    } catch (IOException e) {
      LOG.log(Level.SEVERE, "Failed to write content to file. ", e);
      return false;
    }

    return true;
  }

  public static byte[] readFromFile(String filename) {
    Path path = new File(filename).toPath();
    byte[] res;
    try {
      res = Files.readAllBytes(path);
    } catch (IOException e) {
      LOG.log(Level.SEVERE, "Failed to read from file. ", e);
      res = new byte[0];
    }

    return res;
  }

  public static boolean createDirectory(String directory) {
    return new File(directory).mkdirs();
  }

  public static boolean isDirectoryExists(String directory) {
    return Files.isDirectory(new File(directory).toPath());
  }

  public static boolean isFileExists(String file) {
    return Files.exists(new File(file).toPath());
  }

  public static boolean hasChildren(String file) {
    return isDirectoryExists(file) && new File(file).list().length > 0;
  }

  public static String getBaseName(String file) {
    return new File(file).getName();
  }

  public static boolean deleteDir(String dir) {
    return deleteDir(new File(dir), true);
  }

  public static boolean deleteDir(File dir, boolean deleteSelf) {
    if (Files.isSymbolicLink(dir.toPath())) {
      try {
        Files.delete(dir.toPath());
        return true;
      } catch (IOException e) {
        return false;
      }
    }

    if (dir.isDirectory()) {
      String[] children = dir.list();

      for (String child : children) {
        boolean success = deleteDir(new File(dir, child), true);
        if (!success) {
          return false;
        }
      }
    }

    if (deleteSelf) {
      return dir.delete();
    } else {
      return true;
    }
  }

  public static boolean cleanDir(String dir) {
    return deleteDir(new File(dir), false);
  }
}

File: heron/common/src/cpp/basics/fileutils.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "basics/fileutils.h"
#include <dirent.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include "glog/logging.h"

#include "basics/sprcodes.h"
#include "basics/spconsts.h"

std::string FileUtils::baseName(const std::string& path) {
  return path.substr(path.find_last_of(constPathSeparator) + 1);
}

sp_int32 FileUtils::makeDirectory(const std::string& directory) {
  struct stat st;

  // get the directory name stats
  auto retcode = ::stat(directory.c_str(), &st);

  // check for any errors
  if (retcode == -1) {
    if (errno == ENOENT) {
      if (::mkdir(directory.c_str(), 0700) != 0) {
        PLOG(ERROR) << "Unable to create directory " << directory;
        return SP_NOTOK;
      }
      return SP_OK;
    }

    PLOG(ERROR) << "Unable to create directory " << directory;
    return SP_NOTOK;
  }

  return S_ISDIR(st.st_mode) ? SP_OK : SP_NOTOK;
}

sp_int32 FileUtils::removeFile(const std::string& filepath) {
  if (::unlink(filepath.c_str()) != 0) {
    PLOG(ERROR) << "Unable to delete file " << filepath;
    return SP_NOTOK;
  }
  return SP_OK;
}

sp_int32 FileUtils::removeRecursive(const std::string& directory, bool delete_self) {
  auto dir = ::opendir(directory.c_str());
  if (dir == nullptr) {
    PLOG(ERROR) << "opendir failed for " << directory;
    return SP_NOTOK;
  }

  // list all dir contents
  struct dirent* entry = nullptr;
  while ((entry = ::readdir(dir)) != nullptr) {
    auto current_dir = std::string(entry->d_name).compare(constCurrentDirectory);
    auto parent_dir = std::string(entry->d_name).compare(constParentDirectory);

    if (current_dir != 0 && parent_dir != 0) {
      auto path = directory + constPathSeparator + std::string(entry->d_name);
      if (entry->d_type == DT_DIR) {
        if (removeRecursive(path, true) == SP_NOTOK) return SP_NOTOK;
      } else {
        if (::unlink(path.c_str()) != 0) {
          PLOG(ERROR) << "Unable to delete " << path;
          return SP_NOTOK;
        }
      }
    }
  }

  ::closedir(dir);

  // now delete the dir
  if (delete_self) {
    if (::rmdir(directory.c_str()) != 0) {
      PLOG(ERROR) << "Unable to delete directory " << directory;
      return SP_NOTOK;
    }
  }

  return SP_OK;
}

sp_int32 FileUtils::listFiles(const std::string& directory, std::vector<std::string>& files) {
  auto dir = ::opendir(directory.c_str());
  if (dir != nullptr) {
    struct dirent* ent;
    while ((ent = ::readdir(dir)) != nullptr) {
      auto current_dir = std::string(ent->d_name).compare(constCurrentDirectory);
      auto parent_dir = std::string(ent->d_name).compare(constParentDirectory);
      if (current_dir != 0 && parent_dir != 0) {
        files.push_back(ent->d_name);
      }
    }

    ::closedir(dir);
    return SP_OK;
  }

  // log errors that we could not open the dir
  PLOG(ERROR) << "Unable to open directory " << directory;
  return SP_NOTOK;
}

std::string FileUtils::readAll(const std::string& file) {
  std::ifstream in(file.c_str(), std::ifstream::in | std::ifstream::binary);
  std::stringstream buffer;
  buffer << in.rdbuf();
  return buffer.str();
}

time_t FileUtils::getModifiedTime(const std::string& file) {
  struct stat attrib;
  if (::stat(file.c_str(), &attrib) == 0) {
    return attrib.st_mtime;
  }

  PLOG(ERROR) << "Unable to get file modified time for " << file;
  return SP_NOTOK;
}

bool FileUtils::is_symlink(const std::string& filepath) {
  struct stat attrib;
  if (::lstat(filepath.c_str(), &attrib) == 0) {
    return S_ISLNK(attrib.st_mode);
  }

  PLOG(ERROR) << "Unable to check if file is a symlink " << filepath;
  return false;
}

bool FileUtils::writeAll(const std::string& filename, const char* data, size_t len) {
  std::ofstream ot(filename.c_str(), std::ios::out | std::ios::binary);
  if (!ot) return false;
  ot.write(reinterpret_cast<const char*>(data), len);
  ot.close();
  return true;
}

sp_int32 FileUtils::getCwd(std::string& path) {
  char maxpath[MAXPATHLEN];
  if (::getcwd(maxpath, MAXPATHLEN) == nullptr) {
    PLOG(ERROR) << "Could not get the current working directory";
    return SP_NOTOK;
  }

  path = maxpath;
  return SP_OK;
}
